examples/ssr-data-router/server.js

```diff
@@ -0,0 +1,93 @@
+let path = require("path");
+let fsp = require("fs/promises");
+let express = require("express");
+let { installGlobals } = require("@remix-run/node");
+
+// Polyfill Web Fetch API
+installGlobals();
+
+let root = process.cwd();
+let isProduction = process.env.NODE_ENV === "production";
+
+function resolve(p) {
+  return path.resolve(__dirname, p);
+}
+
+async function createServer() {
+  let app = express();
+  /**
+   * @type {import('vite').ViteDevServer}
+   */
+  let vite;
+
+  if (!isProduction) {
+    vite = await require("vite").createServer({
+      root,
+      server: { middlewareMode: "ssr" },
+    });
+
+    app.use(vite.middlewares);
+  } else {
+    app.use(require("compression")());
+    app.use(express.static(resolve("dist/client")));
+  }
+
+  app.use("*", async (req, res) => {
+    let url = req.originalUrl;
+
+    try {
+      let template;
+      let render;
+
+      if (!isProduction) {
+        template = await fsp.readFile(resolve("index.html"), "utf8");
+        template = await vite.transformIndexHtml(url, template);
+        render = await vite
+          .ssrLoadModule("src/entry.server.tsx")
+          .then((m) => m.render);
+      } else {
+        template = await fsp.readFile(
+          resolve("dist/client/index.html"),
+          "utf8"
+        );
+        render = require(resolve("dist/server/entry.server.js")).render;
+      }
+
+      try {
+        let { hydrationData, html: appHtml } = await render(url);
+        let scriptHtml = `
+          <script>
+            window.__hydrationData = JSON.parse(${JSON.stringify(
+              JSON.stringify(hydrationData)
+            )});
+          </script>
+          `;
```

Comment:
We now get back both the rendered HTML and the `hydrationData` to pass up to the client